edb/pgsql/resolver/command.py

```diff
@@ -339,6 +369,409 @@ def _preprocess_insert_stmt(
+                            ),
+                            compexpr=qlast.Path(
+                                steps=[
+                                    value_ql,
+                                    qlast.Ptr(name=sub_name.name),
+                                    qlast.Ptr(name=ptr_name, type='property'),
+                                ],
+                            ),
+                        )
+                        for ptr_name, _ in value_columns
+                        if ptr_name not in ('source', 'target')
+                    ],
+                ),
+            )
+        ],
+    )
+    if not is_value_single:
+        # value relation might contain multiple rows
+        # to express this in EdgeQL, we must wrap `insert` into a `for` query
+        ql_stmt = qlast.ForQuery(
+            iterator=qlast.Path(steps=[qlast.IRAnchor(name=value_name)]),
+            iterator_alias=iterator_name,
+            result=ql_stmt,
+        )
+    # ql_stmt = qlast.Path(steps=[ql_stmt, qlast.Ptr(name=sub_name.name)])
+
+    ql_returning_shape: List[qlast.ShapeElement] = []
+    if stmt.returning_list:
+        # construct the shape that will extract all needed column of the subject
+        # table (because they might be be used by RETURNING clause)
+        for column in sub_table.columns:
+            if column.hidden:
+                continue
+            if column.name in ('source', 'target'):
+                # no need to include in shape, they will be present anyway
+                continue
+
+            ql_returning_shape.append(
+                qlast.ShapeElement(
+                    expr=qlast.Path(steps=[qlast.Ptr(name=column.name)]),
+                    compexpr=qlast.Path(
+                        partial=True,
+                        steps=[
+                            qlast.Ptr(name=sub_name.name),
+                            qlast.Ptr(name=column.name, type='property'),
+                        ],
+                    ),
+                )
+            )
+
+    return PreprocessedDML(
+        input=stmt,
+        ql_stmt=ql_stmt,
+        ql_returning_shape=ql_returning_shape,
+        ql_singletons={source_id},
+        ql_anchors={value_name: source_id},
+        external_rels={source_id: (value_rel, ('source', 'identity'))},
+        early_result=context.CompiledDML(
+            value_cte_name=value_cte_name,
+            value_relation_input=value_relation,
+            value_columns=value_columns,
+            value_iterator_name=value_iterator,
+            # these will be populated after compilation
+            output_ctes=[],
+            output_relation_name='',
+            output_namespace={},
+        ),
+        # these will be populated after compilation
+        subject_columns=[],
+    )
+
+
+def _preprocess_insert_property_stmt(
+    stmt: pgast.InsertStmt,
+    prop: s_properties.Property,
+    expected_columns: List[context.Column],
+    *,
+    ctx: Context,
+) -> PreprocessedDML:
```

Comment:
I think it should be possible to unify the property and link cases and have it simplify things more than it complicates them?
(Maybe I'm wrong, but probably worth trying for a bit)

Reply:
I've managed to combine link and multi-property functions.